\begin{center}
Gabe Morris
\end{center}

In [1]:
# Notebook Preamble
import sympy as sp
import numpy as np
from IPython.core.interactiveshell import InteractiveShell
import matplotlib.pyplot as plt

InteractiveShell.ast_node_interactivity = 'all'
plt.style.use('maroon_ipynb.mplstyle')

\pagebreak
\tableofcontents
\pagebreak

\begin{center}
\begin{tabular}{c c c}
ME 6233 & Homework 4 & Gabe Morris \\
& & gnm54
\end{tabular}
\end{center}

# Problem 1
## Given

\begin{center}
\includegraphics{images/fig1}
\end{center}

For a rectangular element, the displacements at four nodes are given by

$$u_1=0$$
$$v_1=0$$
$$u_2=-0.5$$
$$v_2=-0.5$$
$$u_3=0.75$$
$$v_3=1.25$$
$$u_4=0.5$$
$$v_4=1$$

## Find
a. Calculate the displacement $(u,v)$ at point $(x,y)=(0.7,1.3)$.
b. Calculate the strain $\epsilon_{xx}$ at point $(x,y)=(0.7,1.3)$

## Solution
From Fig. 3.4-1 in the text, $a=1.5$ and $b=1$. These are the center to edge dimensions of the rectangle and are used in the shape functions/strain displacement matrix.

The analysis of the bilinear quadrilateral depends on the origin being at the center of the rectangle. Therefore, the point $(0.7,1.3)$ relative to the origin at the center is $(0.7,1.3)-(a,b)=(-0.8,0.3)$.

### Part A
The displacements $u,v$ can be defined as the dot product between the shape function and the displacement vectors

$$u(x,y)=\vec{N}\cdot\vec{u}$$
$$v(x,y)=\vec{N}\cdot\vec{v}$$

where the shape function in the vector form is,

$$N=\begin{bmatrix}
\frac{(a - x) (b - y)}{4 a b} \\
\frac{(a + x) (b - y)}{4 a b} \\
\frac{(a + x) (b + y)}{4 a b} \\
\frac{(a - x) (b + y)}{4 a b}
\end{bmatrix}$$

and $\vec{u}$ and $\vec{v}$ are $\langle u_1,u_2,u_3,u_4\rangle$ and $\langle v_1,v_2,v_3,v_4\rangle$.

In [2]:
# Define known parameters
# The underscore denotes a numerical value, while no underscore denotes a symbol.
a_, b_ = 1.5, 1
x_, y_ = -0.8, 0.3
u_, v_ = [0, -0.5, 0.75, 0.5], [0, -0.5, 1.25, 1]
d_ = np.array(list(zip(u_, v_))).flatten()

# Define symbols
a, b, x, y = sp.symbols('a b x y')
u_vec, v_vec = sp.Matrix(u_), sp.Matrix(v_)
d_vec = sp.Matrix(d_)

# Shape function
N = 1/(4*a*b)*sp.Matrix([(a - x)*(b - y), (a + x)*(b - y), (a + x)*(b + y), (a - x)*(b + y)])

sub = [(a, a_), (b, b_), (x, x_), (y, y_)]
N_ = N.subs(sub)

u = sp.DotProduct(N, u_vec)
v = sp.DotProduct(N, v_vec)
sp.Eq(u, N_.dot(u_vec), evaluate=False)
sp.Eq(v, N_.dot(v_vec), evaluate=False)

Eq(DotProduct(Matrix([
[(a - x)*(b - y)/(4*a*b)],
[(a + x)*(b - y)/(4*a*b)],
[(a + x)*(b + y)/(4*a*b)],
[(a - x)*(b + y)/(4*a*b)]]), Matrix([
[   0],
[-0.5],
[0.75],
[ 0.5]])), 0.322083333333333)

Eq(DotProduct(Matrix([
[(a - x)*(b - y)/(4*a*b)],
[(a + x)*(b - y)/(4*a*b)],
[(a + x)*(b + y)/(4*a*b)],
[(a - x)*(b + y)/(4*a*b)]]), Matrix([
[   0],
[-0.5],
[1.25],
[   1]])), 0.647083333333333)

Thus, the displacements $u,v$ at $(-0.8,0.3)$ are $u=0.322$ and $v=0.647$.

### Part B
The strain displacement matrix is

In [3]:
# Getting B
B = 1/(4*a*b)*sp.Matrix([
    [-(b - y), 0, b - y, 0, b + y, 0, -(b + y), 0],
    [0, -(a - x), 0, -(a + x), 0, (a + x), 0, (a - x)],
    [-(a - x), -(b - y), -(a + x), b - y, a + x, b + y, a - x, -(b + y)]
])
sp.Eq(sp.Symbol('B'), B, evaluate=False)

Eq(B, Matrix([
[(-b + y)/(4*a*b),                0,  (b - y)/(4*a*b),                0, (b + y)/(4*a*b),               0, (-b - y)/(4*a*b),                0],
[               0, (-a + x)/(4*a*b),                0, (-a - x)/(4*a*b),               0, (a + x)/(4*a*b),                0,  (a - x)/(4*a*b)],
[(-a + x)/(4*a*b), (-b + y)/(4*a*b), (-a - x)/(4*a*b),  (b - y)/(4*a*b), (a + x)/(4*a*b), (b + y)/(4*a*b),  (a - x)/(4*a*b), (-b - y)/(4*a*b)]]))

The strain may be calculated using $\epsilon=Bd$.

In [4]:
Bd = sp.MatMul(B, d_vec)
Bd_doit = Bd.doit().subs(sub)
sp.Eq(Bd, Bd_doit, evaluate=False)

Eq(Matrix([
[(-b + y)/(4*a*b),                0,  (b - y)/(4*a*b),                0, (b + y)/(4*a*b),               0, (-b - y)/(4*a*b),                0],
[               0, (-a + x)/(4*a*b),                0, (-a - x)/(4*a*b),               0, (a + x)/(4*a*b),                0,  (a - x)/(4*a*b)],
[(-a + x)/(4*a*b), (-b + y)/(4*a*b), (-a - x)/(4*a*b),  (b - y)/(4*a*b), (a + x)/(4*a*b), (b + y)/(4*a*b),  (a - x)/(4*a*b), (-b - y)/(4*a*b)]])*Matrix([
[   0],
[   0],
[-0.5],
[-0.5],
[0.75],
[1.25],
[ 0.5],
[ 1.0]]), Matrix([
[-0.00416666666666666],
[              0.5875],
[   0.333333333333333]]))

Thus, $\epsilon_{xx}=-0.00417$.